### ISSUE TITLE
在训练和测试阶段，batchnorm 的 mean，variance，bias，scale有diff。

### ISSUE THREAD

**Issue by Dely-Yu** [PaddlePaddle/Paddle#19957]
本地，单卡GPU：K40.
fluid1.5
![7475CBAF-8314-49DF-887F-86EEE7DE10E5](https://user-images.githubusercontent.com/29305857/65490490-d7bbe780-dedf-11e9-98c8-5f3e2b7161ea.JPG)
。
数据总量为32张图，训练的batchsize为32，测试batchsize为32.
每个训练iter打出batch层的mean，variance，bias，scale值。
在线测试的时候也输出mean，variance，bias，scale值，想check下测试时batchnorm的参数。

问题是：
1。第100个iter，输出了这4个量，随即测试，也输出了这4个量。发现不一致。预期是应该一致，因为每个训练iter会更新mean，variance，print输出的是更新后的mean，variance。测试的时候直接拿来mean，variance来用。

**Comment by sandyhouse** [comment]
上面的红框对应的是iter100的输出，下面的对应是test的输出吗？有没有可能是打印的时候还没有作为参数更新？

**Comment by Dely-Yu** [comment]
> 上面的红框对应的是iter100的输出，下面的对应是test的输出吗？有没有可能是打印的时候还没有作为参数更新？

1。能不能联系相关同学，给出一个明确的解答
2。我怎样才能精确的check  和 print 每一步mean variance 的更新过程，然后check下infer 时用的mean，variance 。

**Comment by Dely-Yu** [comment]
def ctc_train_net(images, label, mask, args, num_classes,steps):
    regularizer = fluid.regularizer.L2Decay(args.l2)
    T = 25
    #ngram_mask = fluid.layers.data(name="ngram_mask",shape = [T,T])#[b,T,T]
    ngram_label=fluid.layers.data(name="ngram_label",shape=[T,1],dtype = "int64") #[b,T,1]
    src_word_pos=fluid.layers.data(name="src_word_pos",shape=[T,1],dtype = "int64") #[b,T,1]
    word_pos=fluid.layers.data(name="word_pos",shape=[T,1],dtype = "int64") #[b,T,1]
    src_slf_attn_bias1 = fluid.layers.data(name = "src_slf_attn_bias1",shape=[4,T,T])
    src_slf_attn_bias2 = fluid.layers.data(name = "src_slf_attn_bias2",shape=[4,T,T])
    word_out, fc_out,ngram_out  = encoder_net_attdecode(images, num_classes,
                                                src_word_pos ,word_pos,
                                                src_slf_attn_bias1,
                                                src_slf_attn_bias2,
                                                None,#ngram_label,#None,
                                                regularizer=regularizer, gradient_clip=10,n_layer = args.n_layer)
    
    casted_label = fluid.layers.cast(x=label, dtype='int64')
    cost      = fluid.layers.cross_entropy(input=fc_out, label=casted_label)

    casted_label_word = fluid.layers.cast(x=label, dtype='int64')
    cost_word = fluid.layers.cross_entropy(input=word_out, label=casted_label_word)
    
    
    casted_label_word = fluid.layers.cast(x=label, dtype='int64')
    cost_ngram = fluid.layers.cross_entropy(input=ngram_out, label=casted_label_word)
    lbl_weight = fluid.layers.data(name="lbl_weight",shape=[-1,1],dtype='int64')
    
    #fluid.layers.Print(images,message="images",summarize = 50)
    #fluid.layers.Print(fc_out,message="fc_out",summarize = 50)
    #fluid.layers.Print(word_out,message="fc_out",summarize = 50)
    
    #fluid.layers.Print(cost,message="cost",summarize = 50)
    cost = cost * lbl_weight 
    cost_word = cost_word * lbl_weight 
    cost_ngram = cost_ngram * lbl_weight 
    
    img_cost = fluid.layers.reshape(x = fluid.layers.reduce_sum(cost), shape=[1])
    word_cost = fluid.layers.reshape(x = fluid.layers.reduce_sum(cost_word), shape=[1])
    ngram_cost = fluid.layers.reshape(x = fluid.layers.reduce_sum(cost_ngram) , shape=[1]) 
    sum_cost = fluid.layers.sum([word_cost, img_cost,ngram_cost])

    _, decoded_out = fluid.layers.topk(input=fc_out , k=1) 
    decoded_out = fluid.layers.lod_reset(x=decoded_out, y=casted_label)    

    #tmp1 = fluid.layers.control_flow.Print(input = decoded_out)
    #tmp2 = fluid.layers.control_flow.Print(input = casted_label)

    error_evaluator = fluid.evaluator.EditDistance(decoded_out, casted_label)
    node_out = fluid.layers.argmax(fc_out, axis =-1)
    #fluid.layers.Print(casted_label,message="label",summarize = 50)
    #fluid.layers.Print(decoded_out,message="decoded_out")
    #fluid.layers.Print(lbl_weight,message="label_weight",summarize = 50)
    #fluid.layers.Print(node_out,message="node_out",summarize = 50)
    #Block_5_layer2_bn1.w_1 
    Block_5_layer2_bn1_w_1 =fluid.default_main_program().global_block().var( "Block_5_layer2_bn1.w_1" ) #mean
    Block_5_layer2_bn1_w_2 =fluid.default_main_program().global_block().var( "Block_5_layer2_bn1.w_2" ) #variance
    Block_5_layer2_bn1_w_0 =fluid.default_main_program().global_block().var( "Block_5_layer2_bn1.w_0" ) #scale
    Block_5_layer2_bn1_b_0 =fluid.default_main_program().global_block().var( "Block_5_layer2_bn1.b_0" ) #bias

    **fluid.layers.Print(Block_5_layer2_bn1_b_0,message="scale",summarize = 20)
    fluid.layers.Print(Block_5_layer2_bn1_w_0,message="bias",summarize = 20)
    fluid.layers.Print(Block_5_layer2_bn1_w_2,message="mean",summarize = 20)
    fluid.layers.Print(Block_5_layer2_bn1_w_1,message="variance",summarize = 20)**

    #Block_5_layer2_bn1_tmp_0 =fluid.default_main_program().global_block().var( "Block_5_layer2_bn1.tmp_0" ) #saved_mean
    #Block_5_layer2_bn1_tmp_0.persistable = True
    #print(Block_5_layer2_bn1_tmp_0) 
    #Block_5_layer2_bn1_tmp_1 =fluid.default_main_program().global_block().var( "Block_5_layer2_bn1.tmp_1" ) #saved_variance
    #Block_5_layer2_bn1_tmp_1.persistable = True
    #print(Block_5_layer2_bn1_tmp_1) 
    #momentum = 0.9
    #Block_5_layer2_bn1_w_2 = Block_5_layer2_bn1_w_2 * momentum + (1-momentum) * Block_5_layer2_bn1_tmp_1
    #Block_5_layer2_bn1_w_1 = Block_5_layer2_bn1_w_1 * momentum + (1-momentum) * Block_5_layer2_bn1_tmp_0
    #fluid.layers.Print(Block_5_layer2_bn1_w_1,message="Block_5_layer2_bn1.w_1")

    inference_program = fluid.default_main_program().clone(for_test = True)

    if args.learning_rate_decay:
        learning_rate = args.learning_rate
        learning_rate = fluid.layers.piecewise_decay([steps*2, steps*4, steps*5],
                [learning_rate ,learning_rate*0.1, learning_rate * 0.01, learning_rate*0.001])
    else:
        learning_rate = args.learning_rate
    #optimizer = fluid.optimizer.Momentum(learning_rate=args.learning_rate,momentum = 0.9)
    optimizer = fluid.optimizer.Adam(learning_rate= learning_rate)
    _, params_grads = optimizer.minimize(sum_cost)
    model_average = None
    if args.average_window > 0:
        model_average = fluid.optimizer.ModelAverage(
            args.average_window,
            min_average_window=args.min_average_window,
            max_average_window=args.max_average_window)
    return sum_cost, img_cost, word_cost, error_evaluator, inference_program, model_average,node_out

**Comment by sandyhouse** [comment]
把打印放在最后呢，即minimize之后，我怀疑是在更新梯度前调用了打印，所以输出不一致。或者为了排出这种影响，可以在每次run之后打印对应tensor的值对比一下。

**Comment by ceci3** [comment]
用一个包含bn的简单的网络测试： 最后一轮更新过后的参数和预测所用的参数是一致的，可以排查使用问题。
<img width="837" alt="屏幕快照 2019-09-25 上午12 21 20" src="https://user-images.githubusercontent.com/29245900/65530596-dfed4480-df2a-11e9-9878-aaa54bc90648.png">

**Comment by ceci3** [comment]
用一个包含bn的简单的网络测试： 最后一轮更新过后的参数和预测所用的参数是一致的，说明op实现没有问题，需要用户排查bn的使用问题。
<img width="837" alt="屏幕快照 2019-09-25 上午12 21 20" src="https://user-images.githubusercontent.com/29245900/65530630-f0052400-df2a-11e9-8bfa-1c564ac3c30d.png">
